// @mongosh/async-rewriter2/lib/async-writer-babel.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const babel = __importStar(require("@babel/core"));
const runtime_support_nocov_1 = __importDefault(require("./runtime-support.nocov"));
const wrap_as_iife_1 = __importDefault(require("./stages/wrap-as-iife"));
const uncatchable_exceptions_1 = __importDefault(require("./stages/uncatchable-exceptions"));
const transform_maybe_await_1 = __importDefault(require("./stages/transform-maybe-await"));
const error_codes_1 = require("./error-codes");
class AsyncWriter {
    step(code, plugins) {
        var _a;
        return (_a = babel.transformSync(code, {
            plugins,
            code: true,
            configFile: false,
            babelrc: false,
            browserslistConfigFile: false,
            compact: code.length > 10000,
            sourceType: 'script'
        })) === null || _a === void 0 ? void 0 : _a.code;
    }
    process(code) {
        try {
            code = this.step(code, [
                require('@babel/plugin-transform-shorthand-properties').default,
                require('@babel/plugin-transform-parameters').default,
                require('@babel/plugin-transform-destructuring').default
            ]);
            code = this.step(code, [wrap_as_iife_1.default]);
            code = this.step(code, [uncatchable_exceptions_1.default]);
            code = this.step(code, [
                [
                    transform_maybe_await_1.default,
                    { customErrorBuilder: babel.types.identifier('MongoshAsyncWriterError') }
                ]
            ]);
            return code;
        }
        catch (e) {
            e.message = e.message.replace('unknown: ', '');
            throw e;
        }
    }
    runtimeSupportCode() {
        return this.process(`
    class MongoshAsyncWriterError extends Error {
      constructor(message, codeIdentifier) {
        const code = (${JSON.stringify(error_codes_1.AsyncRewriterErrors)})[codeIdentifier];
        super(\`[\${code}] \${message}\`);
        this.code = code;
      }
    }
    ${runtime_support_nocov_1.default}`);
    }
}
exports.default = AsyncWriter;
