// @mongosh/cli-repl/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildInfo = exports.runSmokeTests = exports.getMongocryptdPaths = exports.getStoragePaths = exports.mapCliToDriver = exports.parseCliArgs = exports.CliRepl = exports.MONGOSH_WIKI = exports.TELEMETRY_GREETING_MESSAGE = exports.USAGE = exports.clr = void 0;
const arg_mapper_1 = __importDefault(require("./arg-mapper"));
exports.mapCliToDriver = arg_mapper_1.default;
const arg_parser_1 = require("./arg-parser");
Object.defineProperty(exports, "parseCliArgs", { enumerable: true, get: function () { return arg_parser_1.parseCliArgs; } });
const cli_repl_1 = __importDefault(require("./cli-repl"));
exports.CliRepl = cli_repl_1.default;
const clr_1 = __importDefault(require("./clr"));
exports.clr = clr_1.default;
const config_directory_1 = require("./config-directory");
Object.defineProperty(exports, "getStoragePaths", { enumerable: true, get: function () { return config_directory_1.getStoragePaths; } });
const constants_1 = require("./constants");
Object.defineProperty(exports, "MONGOSH_WIKI", { enumerable: true, get: function () { return constants_1.MONGOSH_WIKI; } });
Object.defineProperty(exports, "TELEMETRY_GREETING_MESSAGE", { enumerable: true, get: function () { return constants_1.TELEMETRY_GREETING_MESSAGE; } });
Object.defineProperty(exports, "USAGE", { enumerable: true, get: function () { return constants_1.USAGE; } });
const mongocryptd_manager_1 = require("./mongocryptd-manager");
Object.defineProperty(exports, "getMongocryptdPaths", { enumerable: true, get: function () { return mongocryptd_manager_1.getMongocryptdPaths; } });
const smoke_tests_1 = require("./smoke-tests");
Object.defineProperty(exports, "runSmokeTests", { enumerable: true, get: function () { return smoke_tests_1.runSmokeTests; } });
const build_info_1 = require("./build-info");
Object.defineProperty(exports, "buildInfo", { enumerable: true, get: function () { return build_info_1.buildInfo; } });
exports.default = cli_repl_1.default;
