// @mongosh/logging/lib/setup-logger-and-telemetry.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupLoggerAndTelemetry = void 0;
const mongodb_redact_1 = __importDefault(require("mongodb-redact"));
const history_1 = require("@mongosh/history");
const util_1 = require("util");
const mongodb_log_writer_1 = require("mongodb-log-writer");
const devtools_connect_1 = require("@mongodb-js/devtools-connect");
class NoopAnalytics {
    identify(_info) { }
    track(_info) { }
}
class MultiSet {
    constructor() {
        this._entries = new Map();
    }
    add(entry) {
        var _a;
        const key = JSON.stringify(Object.entries(entry).sort());
        this._entries.set(key, ((_a = this._entries.get(key)) !== null && _a !== void 0 ? _a : 0) + 1);
    }
    clear() {
        this._entries.clear();
    }
    *[Symbol.iterator]() {
        for (const [key, count] of this._entries) {
            yield [Object.fromEntries(JSON.parse(key)), count];
        }
    }
}
function setupLoggerAndTelemetry(bus, log, makeAnalytics, userTraits, mongosh_version) {
    const { logId } = log;
    let userId;
    let telemetry;
    let analytics = new NoopAnalytics();
    try {
        analytics = makeAnalytics();
    }
    catch (e) {
        log.error('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000001), 'analytics', 'Failed to instantiate analytics provider', e);
    }
    let hasStartedMongoshRepl = false;
    bus.on('mongosh:start-mongosh-repl', (ev) => {
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000002), 'repl', 'Started REPL', ev);
        hasStartedMongoshRepl = true;
    });
    let usesShellOption = false;
    bus.on('mongosh:start-loading-cli-scripts', (event) => {
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000003), 'repl', 'Start loading CLI scripts');
        usesShellOption = event.usesShellOption;
    });
    bus.on('mongosh:connect', function (args) {
        const connectionUri = history_1.redactURICredentials(args.uri);
        const { uri: _uri, ...argsWithoutUri } = args;
        const params = { session_id: logId, userId, connectionUri, ...argsWithoutUri };
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000004), 'connect', 'Connecting to server', params);
        if (telemetry) {
            analytics.track({
                userId,
                event: 'New Connection',
                properties: {
                    mongosh_version,
                    session_id: logId,
                    ...argsWithoutUri
                }
            });
        }
    });
    bus.on('mongosh:new-user', function (id, enableTelemetry) {
        userId = id;
        telemetry = enableTelemetry;
        if (telemetry)
            analytics.identify({ userId, traits: userTraits });
    });
    bus.on('mongosh:update-user', function (id, enableTelemetry) {
        userId = id;
        telemetry = enableTelemetry;
        if (telemetry)
            analytics.identify({ userId, traits: userTraits });
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000005), 'config', 'User updated', { enableTelemetry });
    });
    bus.on('mongosh:error', function (error, context) {
        if (context === 'fatal') {
            log.fatal('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000006), context, `${error.name}: ${error.message}`, error);
        }
        else {
            log.error('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000006), context, `${error.name}: ${error.message}`, error);
        }
        if (telemetry && error.name.includes('Mongosh')) {
            analytics.track({
                userId,
                event: 'Error',
                properties: {
                    mongosh_version,
                    name: error.name,
                    code: error.code,
                    scope: error.scope,
                    metadata: error.metadata
                }
            });
        }
    });
    bus.on('mongosh:evaluate-input', function (args) {
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000007), 'repl', 'Evaluating input', args);
    });
    bus.on('mongosh:use', function (args) {
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000008), 'shell-api', 'Used "use" command', args);
        if (telemetry) {
            analytics.track({
                userId,
                event: 'Use',
                properties: {
                    mongosh_version
                }
            });
        }
    });
    bus.on('mongosh:show', function (args) {
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000009), 'shell-api', 'Used "show" command', args);
        if (telemetry) {
            analytics.track({
                userId,
                event: 'Show',
                properties: {
                    mongosh_version,
                    method: args.method
                }
            });
        }
    });
    bus.on('mongosh:setCtx', function (args) {
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000010), 'shell-api', 'Initialized context', args);
    });
    bus.on('mongosh:api-call-with-arguments', function (args) {
        let arg;
        try {
            arg = JSON.parse(JSON.stringify(args));
        }
        catch (_a) {
            arg = { _inspected: util_1.inspect(args) };
        }
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000011), 'shell-api', 'Performed API call', mongodb_redact_1.default(arg));
    });
    bus.on('mongosh:api-load-file', function (args) {
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000012), 'shell-api', 'Loading file via load()', args);
        if (telemetry) {
            analytics.track({
                userId,
                event: hasStartedMongoshRepl ? 'Script Loaded' : 'Script Loaded CLI',
                properties: {
                    mongosh_version,
                    nested: args.nested,
                    ...(hasStartedMongoshRepl ? {} : { shell: usesShellOption })
                }
            });
        }
    });
    bus.on('mongosh:eval-cli-script', function () {
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000013), 'repl', 'Evaluating script passed on the command line');
        if (telemetry) {
            analytics.track({
                userId,
                event: 'Script Evaluated',
                properties: {
                    mongosh_version,
                    shell: usesShellOption
                }
            });
        }
    });
    bus.on('mongosh:mongoshrc-load', function () {
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000014), 'repl', 'Loading .mongoshrc.js');
        if (telemetry) {
            analytics.track({
                userId,
                event: 'Mongoshrc Loaded',
                properties: {
                    mongosh_version
                }
            });
        }
    });
    bus.on('mongosh:mongoshrc-mongorc-warn', function () {
        log.info('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000015), 'repl', 'Warning about .mongorc.js/.mongoshrc.js mismatch');
        if (telemetry) {
            analytics.track({
                userId,
                event: 'Mongorc Warning',
                properties: {
                    mongosh_version
                }
            });
        }
    });
    bus.on('mongosh:mongocryptd-tryspawn', function (ev) {
        log.info('MONGOCRYPTD', mongodb_log_writer_1.mongoLogId(1000000016), 'mongocryptd', 'Trying to spawn mongocryptd', ev);
    });
    bus.on('mongosh:mongocryptd-error', function (ev) {
        var _a;
        log.warn('MONGOCRYPTD', mongodb_log_writer_1.mongoLogId(1000000017), 'mongocryptd', 'Error running mongocryptd', {
            ...ev,
            error: (_a = ev.error) === null || _a === void 0 ? void 0 : _a.message
        });
    });
    bus.on('mongosh:mongocryptd-log', function (ev) {
        log.info('MONGOCRYPTD', mongodb_log_writer_1.mongoLogId(1000000018), 'mongocryptd', 'mongocryptd log message', ev);
    });
    bus.on('mongosh-snippets:loaded', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000019), 'snippets', 'Loaded snippets', ev);
    });
    bus.on('mongosh-snippets:npm-lookup', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000020), 'snippets', 'Performing npm lookup', ev);
    });
    bus.on('mongosh-snippets:npm-lookup-stopped', function () {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000021), 'snippets', 'npm lookup stopped');
    });
    bus.on('mongosh-snippets:npm-download-failed', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000022), 'snippets', 'npm download failed', ev);
    });
    bus.on('mongosh-snippets:npm-download-active', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000023), 'snippets', 'npm download active', ev);
    });
    bus.on('mongosh-snippets:fetch-index', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000024), 'snippets', 'Fetching snippet index', ev);
    });
    bus.on('mongosh-snippets:fetch-cache-invalid', function () {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000025), 'snippets', 'Snippet cache invalid');
    });
    bus.on('mongosh-snippets:fetch-index-error', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000026), 'snippets', 'Fetching snippet index failed', ev);
    });
    bus.on('mongosh-snippets:fetch-index-done', function () {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000027), 'snippets', 'Fetching snippet index done');
    });
    bus.on('mongosh-snippets:package-json-edit-error', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000028), 'snippets', 'Modifying snippets package.json failed', ev);
    });
    bus.on('mongosh-snippets:spawn-child', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000029), 'snippets', 'Spawning helper', ev);
    });
    bus.on('mongosh-snippets:load-snippet', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000030), 'snippets', 'Loading snippet', ev);
    });
    bus.on('mongosh-snippets:snippet-command', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000031), 'snippets', 'Running snippet command', ev);
        if (telemetry && ev.args[0] === 'install') {
            analytics.track({
                userId,
                event: 'Snippet Install',
                properties: {
                    mongosh_version
                }
            });
        }
    });
    bus.on('mongosh-snippets:transform-error', function (ev) {
        log.info('MONGOSH-SNIPPETS', mongodb_log_writer_1.mongoLogId(1000000032), 'snippets', 'Rewrote error message', ev);
    });
    const deprecatedApiCalls = new MultiSet();
    const apiCalls = new MultiSet();
    bus.on('mongosh:api-call', function (ev) {
        if (ev.deprecated) {
            deprecatedApiCalls.add({ class: ev.class, method: ev.method });
        }
        if (ev.callDepth === 0 && ev.isAsync) {
            apiCalls.add({ class: ev.class, method: ev.method });
        }
    });
    bus.on('mongosh:evaluate-started', function () {
        deprecatedApiCalls.clear();
        apiCalls.clear();
    });
    bus.on('mongosh:evaluate-finished', function () {
        for (const [entry] of deprecatedApiCalls) {
            log.warn('MONGOSH', mongodb_log_writer_1.mongoLogId(1000000033), 'shell-api', 'Deprecated API call', entry);
            if (telemetry) {
                analytics.track({
                    userId,
                    event: 'Deprecated Method',
                    properties: {
                        mongosh_version,
                        ...entry
                    }
                });
            }
        }
        for (const [entry, count] of apiCalls) {
            if (telemetry) {
                analytics.track({
                    userId,
                    event: 'API Call',
                    properties: {
                        mongosh_version,
                        ...entry,
                        count
                    }
                });
            }
        }
        deprecatedApiCalls.clear();
        apiCalls.clear();
    });
    devtools_connect_1.hookLogger(bus, log, 'mongosh', history_1.redactURICredentials);
    bus.on('mongosh-sp:reset-connection-options', function () {
        log.info('MONGOSH-SP', mongodb_log_writer_1.mongoLogId(1000000040), 'connect', 'Reconnect because of changed connection options');
    });
    bus.on('mongosh-editor:run-edit-command', function (ev) {
        log.error('MONGOSH-EDITOR', mongodb_log_writer_1.mongoLogId(1000000045), 'editor', 'Open external editor', mongodb_redact_1.default(ev));
    });
    bus.on('mongosh-editor:read-vscode-extensions-done', function (ev) {
        log.error('MONGOSH-EDITOR', mongodb_log_writer_1.mongoLogId(1000000043), 'editor', 'Reading vscode extensions from file system succeeded', ev);
    });
    bus.on('mongosh-editor:read-vscode-extensions-failed', function (ev) {
        log.error('MONGOSH-EDITOR', mongodb_log_writer_1.mongoLogId(1000000044), 'editor', 'Reading vscode extensions from file system failed', {
            ...ev,
            error: ev.error.message
        });
    });
}
exports.setupLoggerAndTelemetry = setupLoggerAndTelemetry;
